core/src/history.rs

```diff
@@ -22,18 +25,62 @@ impl ExchangeHistory {
             .map(|(event, batch_id)| (event, BatchId(batch_id as _)))
             .collect();
 
-        Ok(ExchangeHistory { _events: events })
+        Ok(ExchangeHistory::from_events(events))
+    }
+
+    fn from_events(events: Vec<(batch_exchange::Event, BatchId)>) -> Self {
+        // NOTE: Since history methods depend on events being sorted, assert
+        // this invariant is true!
+        debug_assert!(events.windows(2).all(|event_pair| {
+            let (_, previous_event_batch) = event_pair[0];
+            let (_, next_event_batch) = event_pair[1];
+            previous_event_batch <= next_event_batch
+        }));
+
+        ExchangeHistory { events }
     }
 
     /// Reads historic exchange events from an `Orderbook` filestore.
     pub fn from_filestore(orderbook_filestore: impl AsRef<Path>) -> Result<Self> {
         ExchangeHistory::read(File::open(orderbook_filestore)?)
     }
+
+    /// Returns an iterator of all events, including their batch IDs, that were
+    /// emitted until the end of the specified batch.
+    fn events_until_end_of_batch(
+        &self,
+        batch_id: impl Into<BatchId>,
+    ) -> impl Iterator<Item = (&batch_exchange::Event, BatchId)> + '_ {
+        let end_batch_id = batch_id.into();
+        self.events
+            .iter()
+            .take_while(move |(_, batch_id)| *batch_id <= end_batch_id)
+            .map(|(event, batch_id)| (event, *batch_id))
+    }
+
+    /// Returns the finalized orderbook at a historic batch.
+    ///
+    /// The finalized orderbook includes all account state changes and orders
+    /// that are considered by the smart contract for solving the specified
+    /// batch.
+    pub fn auction_data_for_batch(
```

Comment:
Couldn't this be expressed via the OrderbookReading trait? I'm still not entirely sure why we cannot make the existing OrderbookReading handle historic batches.

Reply:
:see_no_evil: I totally brainfarted. Yeah that's a super good point, I will refactor this.